### PR TITLE
Store sample zip file with error

### PR DESCRIPTION
**Pull request by patelchandni**
Store .zip file to D:\home\LogFiles\kudu\deployment\BackupErrorZip only if:
- Function app deployment failed with error: Offset to Central Directory cannot be held in an Int64.
- Zip file size is less than 100 MB.
- No .zip file was previously stored in BackupErrorZip to ensure only 1 .zip file will be stored and not be overwritten.

### CODE REVIEW

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 1]
```diff
@@ -869,6 +881,40 @@ private string GetZipDeployPathInfo()
...
+        private void BackupErrorZip(ArtifactDeploymentInfo zipDeploymentInfo)
+        {
+            try
+            {
+                // File created in D:\home\data\SitePackages for current deployment
+                var filePath = Path.Combine(_environment.SitePackagesPath, zipDeploymentInfo.ArtifactFileName);
+
+                // Store only if zip file size < 100 MB
+                long fileBytes = new FileInfo(filePath).Length;
+                float fileMB = (float)fileBytes / (1024 * 1024);
```
Comment (suwatch) [inline Kudu.Services/Deployment/PushDeploymentController.cs:893]: for new codes, start practicing _024 for readability.

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 2]
```diff
@@ -869,6 +881,40 @@ private string GetZipDeployPathInfo()
...
+                    // Make sure D:\home\LogFiles\kudu\deployment\BackupErrorZip exists
+                    var errorDirPath = Path.Combine(_environment.DeploymentTracePath, "BackupErrorZip");
+                    FileSystemHelpers.EnsureDirectory(errorDirPath);
+
+                    // Store only 1 zip file with error
+                    IEnumerable<string> zipFiles = FileSystemHelpers.GetFiles(errorDirPath, "*.zip");
+                    if (zipFiles.Count() == 0)
+                    {
+                        var errorFilePath = Path.Combine(errorDirPath, zipDeploymentInfo.ArtifactFileName);
+                        using (_tracer.Step("{0}: Saving zip file with error to {1} of size {2} MB.", nameof(BackupErrorZip), errorFilePath, fileMB))
```
Comment (suwatch) [inline Kudu.Services/Deployment/PushDeploymentController.cs:905]: make sure this show up on Kudu kusto table as we would need it for chasing this.

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 3]
```diff
@@ -869,6 +881,40 @@ private string GetZipDeployPathInfo()
...
+                float fileMB = (float)fileBytes / (1024 * 1024);
+                if (fileMB < 100)
+                {
+                    // Make sure D:\home\LogFiles\kudu\deployment\BackupErrorZip exists
+                    var errorDirPath = Path.Combine(_environment.DeploymentTracePath, "BackupErrorZip");
+                    FileSystemHelpers.EnsureDirectory(errorDirPath);
+
+                    // Store only 1 zip file with error
+                    IEnumerable<string> zipFiles = FileSystemHelpers.GetFiles(errorDirPath, "*.zip");
+                    if (zipFiles.Count() == 0)
```
[the remark this answers is not in the document]
Comment (patelchandni) [reply Kudu.Services/Deployment/PushDeploymentController.cs]: Thanks. Added this change.

File: Kudu.Services/Deployment/PushDeploymentController.cs (hunk 1 above)
Comment (patelchandni) [reply Kudu.Services/Deployment/PushDeploymentController.cs]: Thanks. Added this change.

**Review by FinVamp1 — APPROVED** [review]
Looks good to me.

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 4]
```diff
@@ -869,6 +881,39 @@ private string GetZipDeployPathInfo()
...
+                // File created in D:\home\data\SitePackages for current deployment
+                var filePath = Path.Combine(_environment.SitePackagesPath, zipDeploymentInfo.ArtifactFileName);
+
+                // Store only if zip file size < 100 MB
+                long fileBytes = new FileInfo(filePath).Length;
+                float fileMB = (float)fileBytes / (1_024 * 1_024);
+                if (fileMB < 100)
+                {
+                    // Make sure D:\home\LogFiles\kudu\deployment\BackupErrorZip exists
+                    var errorDirPath = Path.Combine(_environment.DeploymentTracePath, "BackupErrorZip");
```
Comment (khkh-ms) [inline Kudu.Services/Deployment/PushDeploymentController.cs:897]: Add constant for "BackupErrorZip" in the `Constants.cs` file.

File: Kudu.Services/Deployment/PushDeploymentController.cs (hunk 2 above)
Comment (patelchandni) [reply Kudu.Services/Deployment/PushDeploymentController.cs:904]: Yes, I have tested it. Logs are created in Kusto.

File: Kudu.Services/Deployment/PushDeploymentController.cs [hunk 5]
```diff
@@ -869,6 +881,39 @@ private string GetZipDeployPathInfo()
...
+                    FileSystemHelpers.EnsureDirectory(errorDirPath);
+
+                    // Store only 1 zip file with error
+                    if (FileSystemHelpers.GetFiles(errorDirPath, "*.zip").Length == 0)
+                    {
+                        var errorFilePath = Path.Combine(errorDirPath, zipDeploymentInfo.ArtifactFileName);
+                        using (_tracer.Step("{0}: Saving zip file with error to {1} of size {2} MB.", nameof(BackupErrorZip), errorFilePath, fileMB))
+                        {
+                            File.Copy(filePath, errorFilePath);
```
Comment (khkh-ms) [inline Kudu.Services/Deployment/PushDeploymentController.cs:906]: Probably use `FileSystemHelpers.CopyFile` instead of `File.Copy` for better testability.

**Pull request merged**